src/hotspot/cpu/x86/gc/g1/g1_x86_64.ad

```diff
@@ -0,0 +1,370 @@
+                         RegSet::of($res$$Register) /* no_preserve */);
+    __ movq($tmp1$$Register, $newval$$Register);
+    __ lock();
+    __ cmpxchgq($tmp1$$Register, Address($mem$$Register, 0));
+    __ setb(Assembler::equal, $res$$Register);
+    __ movzbl($res$$Register, $res$$Register);
+    g1_post_write_barrier(masm, this,
+                          $mem$$Register /* store_addr */,
+                          $tmp1$$Register /* new_val */,
+                          $tmp2$$Register /* tmp1 */,
+                          $tmp3$$Register /* tmp2 */);
+  %}
+  ins_pipe(pipe_cmpxchg);
+%}
+
+instruct g1CompareAndSwapN(rRegI res, indirect mem, rRegN newval, rRegP tmp1, rRegP tmp2, rRegP tmp3, rax_RegN oldval, rFlagsReg cr)
+%{
+  predicate(UseG1GC && n->as_LoadStore()->barrier_data() != 0);
+  match(Set res (CompareAndSwapN mem (Binary oldval newval)));
+  match(Set res (WeakCompareAndSwapN mem (Binary oldval newval)));
+  effect(TEMP res, TEMP tmp1, TEMP tmp2, TEMP tmp3, KILL oldval, KILL cr);
+  format %{ "lock\n\t"
+            "cmpxchgq $newval, $mem\n\t"
+            "sete     $res\n\t"
+            "movzbl   $res, $res" %}
+  ins_encode %{
+    assert_different_registers($oldval$$Register, $mem$$Register);
+    g1_pre_write_barrier(masm, this,
+                         $mem$$Register /* obj */,
+                         $tmp2$$Register /* pre_val */,
+                         $tmp3$$Register /* tmp */,
+                         RegSet::of($mem$$Register, $newval$$Register, $oldval$$Register) /* preserve */,
+                         RegSet::of($res$$Register) /* no_preserve */);
+    __ movl($tmp1$$Register, $newval$$Register);
+    __ lock();
+    __ cmpxchgl($tmp1$$Register, Address($mem$$Register, 0));
+    __ setb(Assembler::equal, $res$$Register);
+    __ movzbl($res$$Register, $res$$Register);
+    __ decode_heap_oop($tmp1$$Register);
+    g1_post_write_barrier(masm, this,
+                          $mem$$Register /* store_addr */,
+                          $tmp1$$Register /* new_val */,
+                          $tmp2$$Register /* tmp1 */,
+                          $tmp3$$Register /* tmp2 */);
+  %}
+  ins_pipe(pipe_cmpxchg);
+%}
+
+instruct g1XChgP(indirect mem, rRegP newval, rRegP tmp1, rRegP tmp2, rRegP tmp3, rFlagsReg cr)
+%{
+  predicate(UseG1GC && n->as_LoadStore()->barrier_data() != 0);
+  match(Set newval (GetAndSetP mem newval));
+  effect(TEMP tmp1, TEMP tmp2, TEMP tmp3, KILL cr);
+  format %{ "xchgq    $newval, $mem" %}
+  ins_encode %{
+    assert_different_registers($mem$$Register, $newval$$Register);
+    g1_pre_write_barrier(masm, this,
+                         $mem$$Register /* obj */,
+                         $tmp2$$Register /* pre_val */,
+                         $tmp3$$Register /* tmp */,
+                         RegSet::of($mem$$Register, $newval$$Register) /* preserve */);
+    __ movq($tmp1$$Register, $newval$$Register);
+    __ xchgq($newval$$Register, Address($mem$$Register, 0));
```

Comment:
Optimization idea: Despite its name, `g1_pre_write_barrier` can be moved after the xchg operation because there's no safepoint within this MachNode. This allows avoiding loading the old value twice.

Reply:
Thanks, I also prototyped this [here](https://github.com/robcasloz/jdk/blob/JDK-8334060-g1-late-barrier-expansion-x64-optimizations/src/hotspot/cpu/x86/gc/g1/g1_x86_64.ad) (guarded by `UseExchangedValueinPreBarriers`). The atomic barrier implementation in this PR is purposefully simple based on 1) the assumption that the cost of the atomic operations will dominate that of their barriers, and 2) the risk of introducing subtle bugs which can be difficult to catch by regular testing. Because of this, I feel hesitant to introduce this kind of optimizations for atomic operation barriers. But I am happy to reconsider, if you have any specific benchmark/configuration in mind where the benefit could outweigh the cost.

Reply:
Note that we had such an optimization already in C2: https://github.com/openjdk/jdk8u-dev/blob/4106121e0ae42d644e45c6eab9037874110ed670/hotspot/src/share/vm/opto/library_call.cpp#L3114
But, it's probably not a big deal. Maybe I can try it on PPC64 which may be more sensitive to accesses on contended memory.

Reply:
Thanks for the reference, I would still prefer to keep this part as is for simplicity. We can always optimize the atomic barriers in follow-up work, if a need arises.

Reply:
After thinking more about this, I figured out that we can optimize more when moving the pre_barrier after the cmpxchg. We can skip all G1 barriers if the cmpxchg fails:
https://github.com/TheRealMDoerr/jdk/blob/a48598075862f17e7b1cfbec29af4c2431809257/src/hotspot/cpu/ppc/gc/g1/g1_ppc.ad#L171
This may reduce load on GC queue handling and related work for GC threads. I'm testing this version and I actually like it more than the version I had before. Please take a look.

(Note that my final version will need https://github.com/openjdk/jdk/pull/20689 to be integrated and merged into your PR.)

Reply:
Right, that makes sense since for PPC's cmpxchg implementation (unlike x64 or aarch64+LSE) you are already explicitly branching on failure anyway.